File: scsc-monolith/src/main/java/demo/scsc/commandside/order/ProductValidation.java
package demo.scsc.commandside.order;

import demo.scsc.Constants;
import demo.scsc.api.productcatalog.ProductUpdateReceivedEvent;
import demo.scsc.config.JpaPersistenceUnit;
import jakarta.persistence.EntityManager;
import org.axonframework.config.ProcessingGroup;
import org.axonframework.eventhandling.EventHandler;
import org.jetbrains.annotations.NotNull;

import java.math.BigDecimal;
import java.util.UUID;

@ProcessingGroup(Constants.PROCESSING_GROUP_PRODUCT)
public class ProductValidation {

    @EventHandler
    public void on(ProductUpdateReceivedEvent productUpdateReceivedEvent) {
        EntityManager em = JpaPersistenceUnit.forName("SCSC").getNewEntityManager();
        em.getTransaction().begin();
        em.merge(toEntity(productUpdateReceivedEvent));
        em.getTransaction().commit();
    }

    public ProductValidationInfo forProduct (UUID id) {

        ProductValidationInfo productValidationInfo = null;

        EntityManager em = JpaPersistenceUnit.forName("SCSC").getNewEntityManager();
        ProductValidationEntity productValidationEntity = em.find(ProductValidationEntity.class, id);
        if (productValidationEntity != null) {
            productValidationInfo = new ProductValidationInfo(productValidationEntity);
        }

        return productValidationInfo;
    }

    @NotNull
    private ProductValidationEntity toEntity(@NotNull ProductUpdateReceivedEvent productUpdateReceivedEvent) {
        ProductValidationEntity productValidationEntity = new ProductValidationEntity();
        productValidationEntity.setId(productUpdateReceivedEvent.id());
        productValidationEntity.setName(productUpdateReceivedEvent.name());
        productValidationEntity.setPrice(productUpdateReceivedEvent.price());
        productValidationEntity.setOnSale(productUpdateReceivedEvent.onSale());
        return productValidationEntity;
    }

    public static class ProductValidationInfo {

        public final String name;
        public final BigDecimal price;
        public final boolean onSale;

        public ProductValidationInfo (ProductValidationEntity productValidationEntity) {
            this.name = productValidationEntity.getName();
            this.price = productValidationEntity.getPrice();
            this.onSale = productValidationEntity.isOnSale();
        }
    }
}

File: scsc-monolith/src/main/kotlin/demo/scsc/commandside/order/ProductValidation.kt
package demo.scsc.commandside.order

import com.typesafe.config.Config
import demo.scsc.Constants
import demo.scsc.api.productcatalog
import demo.scsc.config.resolver.ValidatorFactory
import demo.scsc.util.tx
import org.axonframework.config.ProcessingGroup
import org.axonframework.eventhandling.EventHandler
import java.math.BigDecimal
import java.util.UUID

@ProcessingGroup(Constants.PROCESSING_GROUP_PRODUCT)
class ProductValidation(private val appConfig: Config): ValidatorFactory<UUID, ProductValidation.ProductValidationInfo?> {

    @EventHandler
    fun on(event: productcatalog.ProductUpdateReceivedEvent) {
        tx(appConfig) { it.merge(event.toEntity()) }
    }

    private fun productcatalog.ProductUpdateReceivedEvent.toEntity() = Product(
        id = id,
        name = name,
        price = price,
        isOnSale = onSale
    )

    class ProductValidationInfo(product: Product) {
        val name: String = product.name
        val price: BigDecimal = product.price // TODO switch to monetary amount
        val forSale: Boolean = product.isOnSale
    }

    override fun validator(subject: UUID): ProductValidationInfo? =
         tx(appConfig) {
            it.find(Product::class.java, subject)?.let { product -> ProductValidationInfo(product) }
    }
}
